File: python/isbn-verifier/isbn_verifier.py
def is_valid(isbn):
    digits = [ch if ch != 'X' else '10' for ch in isbn if ch != '-']

    if len(digits) != 10 or '10' in digits[:9] or any(not n.isdigit() for n in digits):
        return False

    return sum(int(n) * i for n, i in zip(digits, range(10, 0, -1))) % 11 == 0

File: haskell/isbn-verifier/src/IsbnVerifier.hs
module IsbnVerifier
  ( isbn
  ) where

import           Data.Char                      ( isDigit )

isbn :: String -> Bool
isbn xs = isValidFormat && isValidSum
 where
  digits  = map (\c -> if c /= 'X' then [c] else "10") (filter (/= '-') xs)
  digits' = concat digits
  isValidFormat =
    length digits == 10 && "10" `notElem` take 9 digits && all isDigit digits'
  isValidSum =
    sum (map (\(x, y) -> read x * y) (zip digits [10, 9 .. 0])) `mod` 11 == 0
